### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by VarLad** [withbayes/Tapir.jl#140]
Thanks for this! 
While using this with with DifferentiationInterface, I get worse performance numbers. Thought reporting it here would make sense.

![image](https://github.com/withbayes/Tapir.jl/assets/51269425/f4ec06a7-d122-4e09-b165-5a1c7f90e324)

Do tell if its a DifferentiationInterface.jl issue.

**Comment by willtebbutt** [comment]
Thanks for opening this @VarLad -- there's an awfully high compile time here. @gdalle I suspect this isn't quite how the DI interface is meant to be used --could you comment?

**Comment by CarloLucibello** [comment]
I think a fair comparison should include the preparation time as well.

**Comment by gdalle** [comment]
Same debate as whether or not we include compilation time when benchmarking Julia against other languages.

**Comment by gdalle** [comment]
> Unfortunately, on my laptop I still see huge compilation time when I do that, so I would guess it's up to Tapir then

Note that when I change the function to something that will be bad for Zygote, like

```julia
function f(x)
    s = zero(eltype(x))
    for i in eachindex(x)
        s += sin(x[i])
    end
    return s
end
```

then my (biased) benchmarking shows Tapir clearly winning again. So it must have something to do with how the `rrule` is computed / used for

```julia
f(x) = sqrt(sum(abs2, x) * sum(abs2, x))
```

**Comment by VarLad** [comment]
I can reproduce the same. There's something wrong with the `sum` operator and Tapir. 
And this shows in both the code with preparation and without.
Although, even with the provided example, Enzyme's results show that there's a lot of performance left to be optimized for :eyes: 

![image](https://github.com/withbayes/Tapir.jl/assets/51269425/7d28bc4a-5ce0-4fe6-a744-62f3ddb87e6e)

**Comment by gdalle** [comment]
I'm very puzzled by this behavior of Chairmarks. Normally the setup code shouldn't be included in the measurements. Maybe @LilithHafner can help shed some light?

**Comment by willtebbutt** [comment]
Ohh there's a difference between the performance of Tapir the first time you run it, and the the second time you run it. The first time you run it, a bunch of allocations happen (various stacks get allocated etc).

The following ought to work:
```julia
]add Chairmarks DifferentiationInterface Tapir Zygote
using Chairmarks, DifferentiationInterface
import Tapir, Zygote

f(x) = sqrt(sum(abs2, x) * sum(abs2, x))

function prep_grad(f, ::AutoTapir, x)
    extras = prepare_gradient(f, AutoTapir(), rand(10000)) 
    value_and_gradient(f, AutoTapir(), x, extras)
    return extras
end

@b (
    x=rand(10000),
    extras=prep_grad(f, AutoTapir(), rand(10000)) ,
) value_and_gradient(f, AutoTapir(), _.x, _.extras) evals=1
```

So it _is_ a Tapir.jl thing, but not quite what we thought. The runtime performance is good, just not great the first time you run it.

**Comment by gdalle** [comment]
> So it is a Tapir.jl thing, but not quite what we thought. The runtime performance is good, just not great the first time you run it.

Presumably the overhead on the first run _even after `build_rrule`_ is still big enough that Chairmarks doesn't bother running the code several times, and just returns that one run.

> edit: this issue does raise the question of how best to communicate the correct way to benchmark Tapir performance. 

Well, in the spirit of "prepare once, run a thousand times" (which is the one adopted by DI), I guess we should add one call to the operator inside the preparation workflow?

**Comment by willtebbutt** [comment]
> Well, in the spirit of "prepare once, run a thousand times" (which is the one adopted by DI), I guess we should add one call to the operator inside the preparation workflow?

I think this is probably the right thing to do -- if it causes problems for some reason, then we can always revert it at a later point in time.

**Comment by gdalle** [comment]
Can you help out with that one? https://github.com/gdalle/DifferentiationInterface.jl/pull/254

**Comment by willtebbutt** [comment]
I _believe_ that if you follow @gdalle 's original approach to benchmarking [here](https://github.com/withbayes/Tapir.jl/issues/140#issuecomment-2106206118) you should find that it works nicely now. If you could confirm that it indeed performs as expected for you, I'll close this issue.

**Comment by gdalle** [comment]
And if you use the main branch of DifferentiationInterface. Release coming tomorrow or so

**Comment by gdalle** [comment]
@willtebbutt it might be worth mentioning the preparation step in the readme when you discuss the DI bindings?

**Comment by LilithHafner** [comment]
> Presumably the overhead on the first run even after `build_rrule` is still big enough that Chairmarks doesn't bother running the code several times, and just returns that one run.

This is correct. Chairmarks and BenchmarkTools both make the (in this case false) assumption that sample runtime is equal to evaluation runtime times number of evaluations, and reports the extrapolated evaluation runtime. I'll look into providing better results in this case.

**Comment by willtebbutt** [comment]
```julia
]add Chairmarks DifferentiationInterface Tapir Zygote
using Chairmarks, DifferentiationInterface
import Tapir, Zygote

f(x) = sqrt(sum(abs2, x) * sum(abs2, x))

@b (
    x=rand(10000),
    extras=prepare_gradient(sum, AutoTapir(safe_mode=false), rand(10000))
) value_and_gradient(sum, AutoTapir(), _.x, _.extras)
```
I just ran the above and got
```julia
94.448 μs (2 allocs: 78.172 KiB)
```
which is what I would expect. I'm using versions:
```julia
  [0ca39b1e] Chairmarks v1.2.1
  [a0c0ee7d] DifferentiationInterface v0.4.0
  [07d77754] Tapir v0.2.12
  [e88e6eb3] Zygote v0.6.70
```

Consequently I'm going to mark this as resolved. Thanks all for contributing to this -- it's been very useful to iron this problem out!
